lib/ftp/src/ftp.erl

```diff
@@ -1285,7 +1287,7 @@ handle_info({Trpt, Socket, Data},
 
 handle_info({Trpt, Socket, Data}, #state{dsock = {Trpt,Socket}, client = From,
                                         caller = recv_chunk}
-            = State) when Trpt==tcp;Trpt==ssl ->
+            = State) when (Trpt==tcp orelse Trpt==ssl) andalso (From =/= undefined) ->
```

Comment:
What value does this defensive approach add?  It is a bug if From is undefined here and now you will get a function clause on handle_info instead of gen_server:reply which is less informative for finding the bug!

Reply:
The comment is relevant for several of the changes below!

Reply:
I completely agree but from the point of view of a static analysis tool, if `From` can be `undefined` because it is not excluded, so eqWAlizer and any other type checker will report this as an error, because it knows that `gen_server:reply` cannot accept an `undefined`. 

I think it depends on the approach that we want to take to make type checkers happy :)

Reply:
I am torn on this one. 

I agree that you do get the error on `gen_server:reply` before, and now you get it in the function clause.

At the same time, if there is an error with an input argument, maybe you want to crash early. 

In this case, it is obvious that the call to `gen_server:reply` crashes. However, if the `From` value were to travel to a bunch of other functions until reaching the `gen_server:reply` call, I do wonder: isn't it more useful to know the first place that should not have received the unexpected value? It seems like the closer you get to the root cause, the better?

Reply:
I do not feel I like to make my code defensive to make a tool happy. I think it is better to let it crash whenever the parameter is used because then you immediately know which parameter has an unexpected value. Typically state machine/gen server function clause will match several values and you will first have to determine which value was it unhappy with and then you can start analyzing when did it become wrong. As it is the state that has an unexpected value the fault is somewhere else than in the function clause that failed in the defensive version. It will be in another previously executed function clause where the returned state was not properly populated.

Reply:
I got rid of the `From =/= undefined` :+1: